tests/debugging/test_encoding.py

```diff
@@ -0,0 +1,284 @@
+        assert_args({"a", "kwargs"})
+        assert_locals(set())
+
+    def arg_and_args_and_kwargs(a, *ars, **kwars):
+        assert_args({"a", "ars", "kwars"})
+        assert_locals(set())
+
+    def args_and_kwargs(*ars, **kwars):
+        assert_args({"ars", "kwars"})
+        assert_locals(set())
+
+    def args(*ars):
+        assert_args({"ars"})
+        assert_locals(set())
+
+    arg_and_kwargs(1, b=2)
+    arg_and_args_and_kwargs(1, 42, b=2)
+    args_and_kwargs()
+    args()
+
+
+@pytest.mark.parametrize(
+    "value,serialized",
+    [
+        # Simple
+        ("foo", "'foo'"),
+        (10, "10"),
+        (0.2, "0.2"),
+        (True, "True"),
+        (None, "None"),
+        (b"Hello", "b'Hello'"[PY2:]),
+        # Container
+        ({"Hello": "World"}, "{'Hello': 'World'}"),
+        ({"Hello": 42}, "{'Hello': 42}"),
+        ({42: "World"}, "{42: 'World'}"),
+        ({}, "{}"),
+        (["Hello", 42, True, None, 10.0], "['Hello', 42, True, None, 10.0]"),
+        ([], "[]"),
+        (tuple(), "()"),
+        ((None,), "(None)"),
+        (set(), "set()"),
+        ({0.1}, "{0.1}"),
+        (
+            ({"Hello": [None, 42, True, None, {b"World"}, 0.07]},),
+            "({'Hello': [None, 42, True, None, {b'World'}, 0.07]})"
+            if PY3
+            else "({'Hello': [None, 42, True, None, {'World'}, 0.07]})",
+        ),
+    ],
+)
+def test_serialize(value, serialized):
+    assert _serialize(value, level=-1) == serialized
+
+
+def test_serialize_custom_object():
+
+    assert _serialize(Custom(), level=-1) == (
+        "Custom(some_arg=({'Hello': [None, 42, True, None, {b'World'}, 0.07]}))"
+        if PY3
+        else "Custom(some_arg=({'Hello': [None, 42, True, None, {'World'}, 0.07]}))"
+    )
+
+    q = "class" if PY3 else "type"
+    assert _serialize(Custom()) == "Custom(some_arg=<%s 'tuple'>)" % q
+    assert _serialize(Custom(), 2) == "Custom(some_arg=(<%s 'dict'>))" % q
+    assert _serialize(Custom(), 3) == "Custom(some_arg=({'Hello': <%s 'list'>}))" % q
+    assert _serialize(Custom(), 4) == "Custom(some_arg=({'Hello': [None, 42, True, None, <%s 'set'>, 0.07]}))" % q
+
+    assert _serialize(Custom) == repr(Custom)
+
+
+@pytest.mark.parametrize(
+    "value,serialized",
+    [
+        (list(range(20)), "[" + ", ".join(map(str, range(10))) + ", ...]"),
+        (tuple(range(20)), "(" + ", ".join(map(str, range(10))) + ", ...)"),
+        (set(range(20)), "{" + ", ".join(map(str, range(10))) + ", ...}"),
+        (list(range(10)), "[" + ", ".join(map(str, range(10))) + "]"),
+        (tuple(range(10)), "(" + ", ".join(map(str, range(10))) + ")"),
+        (set(range(10)), "{" + ", ".join(map(str, range(10))) + "}"),
+    ],
+)
+def test_serialize_collection_max_size(value, serialized):
+    assert _serialize(value) == serialized
+
+
+def test_serialize_long_string():
+    assert _serialize("x" * 11, max_str_len=10) == repr("x" * 9 + "...")
+
+
+def test_captured_context_default_level():
```

Comment:
Could you explain a bit about what  `_captured_value_v2()` and `get_fields()` are doing on a larger scale? I'm getting a bit lost at that level, perhaps because I don't know much about encoding... or because I don't know much about the debugger.

Reply:
`_captured_value_v2` serialises a Python object into a `dict` with keys `type` and `value`. If the object is not a "scalar" value (e.g. `int`, `float`, or even `str`), we also add `fields` in which we serialise the object's fields in a recursive fashion.

`get_fields` is the helper that we use to get the fields from an arbitrary Python object. We need to take care of the fact that accessing fields of an object might cause side-effects (e.g. a `__getattribute__` that triggers DB calls or some other form of I/O), so we have to do a bit of extra work in that case.